### PR TITLE
fix(#2766): Virtualizer does not scroll items into view within the browser viewport

### LINKED ISSUE #2766: Collections do not scroll items into view within the browser viewport with keyboard
# 🐛 Bug Report 

When the bounds of a virtualized collection extend beyond the browser viewport, and setting focus to an item within the collection that is out of view, the browser will not scroll the focused item into view.

## 🤔 Expected Behavior

An item receiving focus in a virtualized collection should not only scroll the item into view within the collection, but it should scroll the item into view within the browser as well.

## 😯 Current Behavior

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/0661785f30971bdbcbd291282b26836cf272c3f8/storybook/iframe.html?id=listbox--async-loading&viewMode=story
2. Reduce the inner height of the browser window to be less than the height of the ListBox within the story,  ~200px should be good.
3. Tab to set keyboard focus to the first Item within the ListBox.
4. Use <kbd>ArrowDown</kbd> to scroll until you focus an Item within the ListBox that is outside of the browser viewport.
5. Notice that the browser does not automatically scroll the `document.body` to bring the focused Item into view.

## 💁 Possible Solution

Add the following to `useVirtualizer` method in `@react-aria/virtualizer/src/Virtualizer.tsx`:

```tsx
  // Define an IntersectionObserver to evaluate whether the browser should scroll the element receiving focus into view.
  let intersectionObserver = useMemo(() => {
    const intersectionObserverOptions:IntersectionObserverInit = {
      root: undefined,
      rootMargin: '0px', 
      threshold: 1
    };

    const scrollIntoViewOptions:ScrollIntoViewOptions = {
      behavior: 'auto',
      block: 'nearest',
      inline: 'nearest'
    };

    const intersectionObserverCallback = (entries:Array<IntersectionObserverEntry>, observer:IntersectionObserver) => {
      entries.forEach(entry => {
        if (!entry.isIntersecting) {
          entry.target.scrollIntoView(scrollIntoViewOptions);
        }
        observer.unobserve(entry.target);
      });
    };

    return new IntersectionObserver(intersectionObserverCallback, intersectionObserverOptions);
  }, []);
```
Then at the end of the `onFocus` method, add the following after updating `isFocusWithin.current`:

```tsx
    // Evaluate whether the browser should scroll to bring element receiving focus into view.
    if (isFocusWithin.current) {
      intersectionObserver.observe(e.target);
    }
```

## 🔦 Context

This bug was reported by a team using a static TableView that noticed when the TableView was larger than the browser viewport, focusing a row would not cause the browser to scroll it into view.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  @react-aria/virtualizer@3.3.5
| Browser          |  Chrome Version 97.0.4692.71 (Official Build) (x86_64)
| Operating System | macOS Version 12.1 (21C52)

## 🧢 Your Company/Team

Adobe/Accessibility Adobe/Admin_Console

## 🕷 Tracking Issue (optional)

Jira issue A11Y-5009 and TRON-8092

### PR DESCRIPTION
**Pull request by majornista**
Add IntersectionObserver to detect whether or not a focused item is visible within the browser window, and if not scroll it into view.

Closes #2766 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue #2766.
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/d5eb632b807c8dda8faea50e3cbcc5995528d6bf/storybook/index.html?path=/story/listbox--async-loading&id=listbox--async-loading&providerSwitcher-toastPosition=bottom&viewMode=story
2. Reduce the inner height of the browser window to be less than the height of the ListBox within the story, ~200px should be good.
3. <kbd>Tab</kbd> to set keyboard focus to the first Item within the ListBox.
4. Use <kbd>ArrowDown</kbd> to scroll until you focus an Item within the ListBox that is outside of the browser viewport.
5. Notice that the browser will now automatically scroll the `document.body` to bring the focused Item into view.

Other stories to test: 

- https://reactspectrum.blob.core.windows.net/reactspectrum/d5eb632b807c8dda8faea50e3cbcc5995528d6bf/storybook/index.html?path=/story/tableview--async-loading&id=tableview--async-loading&providerSwitcher-toastPosition=bottom&viewMode=story
- https://reactspectrum.blob.core.windows.net/reactspectrum/d5eb632b807c8dda8faea50e3cbcc5995528d6bf/storybook/index.html?path=/story/useselectablelist--static-ul-static-sub-ul&id=useselectablelist--static-ul-static-sub-ul&providerSwitcher-toastPosition=bottom&viewMode=story
- https://reactspectrum.blob.core.windows.net/reactspectrum/d5eb632b807c8dda8faea50e3cbcc5995528d6bf/storybook/index.html?path=/story/focusscope--focus-safely&id=focusscope--focus-safely&providerSwitcher-toastPosition=bottom&viewMode=story

## 🧢 Your Project:
Adobe/Accessibility Adobe/Admin_Console Jira issues A11Y-5009 and TRON-8092

### CODE REVIEW

**Comment by devongovett** [comment]
This issue is not specific to Virtualizer. For example, use your same reproduction steps on [this example](https://reactspectrum.blob.core.windows.net/reactspectrum/0661785f30971bdbcbd291282b26836cf272c3f8/storybook/iframe.html?id=useselectablelist--static-ul-static-sub-ul&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&providerSwitcher-toastPosition=bottom&viewMode=story).

The root cause is that we use `focusSafely` [here](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/selection/src/useSelectableItem.ts#L117), which prevents scroll. That is very much intentional (e.g. #317). We currently have another bug where we scroll into view where we don't want to (when tapping a card). I'm not sure that scrolling into view always the behavior we want. I can think of cases e.g. involving popovers where this would cause the popover to close, for example. Will need to think about it more.

File: packages/@react-spectrum/card/src/CardView.tsx [hunk 1]
```diff
@@ -195,6 +196,20 @@ function InternalCard(props) {
     focusMode: 'cell'
   }, state, unwrappedRef);
 
+  // Prevent space key from scrolling the CardView if triggered on a disabled item or on a Card in a selectionMode="none" CardView.
+  let allowsInteraction = state.selectionManager.selectionMode !== 'none';
+  let isDisabled = !allowsInteraction || state.disabledKeys.has(item.key);
```
Comment (LFDanLu) [inline packages/@react-spectrum/card/src/CardView.tsx:201]: On second thought, might want to remove all this card specific stuff from my other PR so the scope for this PR is more defined

**Review by LFDanLu — COMMENTED** [review]
This seems to work well enough. I'm a bit concerned about how the scrolling may close dropdowns/popovers which could pose a problem for react-aria users who use `useOverlayPosition`, specifically if they have something like `shouldFlip: false` combine with a positioning that would render the dropdown/popover offscreen (reproducible by modifying the `useOverlayTrigger` docs example with `shouldFlip:false` and `position: bottom`), perhaps edge casey? Our components sidestep this since the overlays don't render offscreen

**Comment by majornista** [comment]
> This seems to work well enough. I'm a bit concerned about how the scrolling may close dropdowns/popovers which could pose a problem for react-aria users who use `useOverlayPosition`, specifically if they have something like `shouldFlip: false` combined with a positioning that would render the dropdown/popover offscreen (reproducible by modifying the `useOverlayTrigger` docs example with `shouldFlip:false` and `position: bottom`), perhaps edge casey? Our components sidestep this since the overlays don't render offscreen

Hmmm... The behavior one can reproduce adding `shouldFlip:false` and `position: bottom` to the `useOverlayPosition` configuration for an overlay trigger near the bottom of the page seems to be a problem with `useOverlayPosition` and `useCloseOnScroll`. We should either not close a popover on scroll when `shouldFlip:false`, or force flipping when space is not available to render the overlay completely without scrolling. 

We will have accessibility problems with any popovers that cannot fit with the available viewport if there is no way to scroll the document body to read the popover content that does not fit without closing the popover.

**Comment by majornista** [comment]
> This issue is not specific to Virtualizer. For example, use your same reproduction steps on [this example](https://reactspectrum.blob.core.windows.net/reactspectrum/0661785f30971bdbcbd291282b26836cf272c3f8/storybook/iframe.html?id=useselectablelist--static-ul-static-sub-ul&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&providerSwitcher-toastPosition=bottom&viewMode=story).
> 
> The root cause is that we use `focusSafely` [here](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/selection/src/useSelectableItem.ts#L117), which prevents scroll. That is very much intentional (e.g. #317). We currently have another bug where we scroll into view where we don't want to (when tapping a card). I'm not sure that scrolling into view always the behavior we want. I can think of cases e.g. involving popovers where this would cause the popover to close, for example. Will need to think about it more.

I think #317 is treating a symptom rather than the underlying issue. `useCloseOnScroll` was added to avoid having to reposition an overlay rendered to a portal on scroll. I don't think this is the correct approach. For accessibility, a user should be able to scroll the content of an overlay that extends beyond the viewport into view without closing the overlay. With `useCloseOnScroll`, as currently implemented, this is not possible.

**Comment by devongovett** [comment]
I don't follow. Popovers should never extend beyond the viewport. The content of the popover itself is scrollable, not the document body.

Regardless, closing on scroll has very little to do with this and it is far from the only reason for the current behavior. In fact, this is the first I’m hearing of it in relation to this PR. We used the `preventScroll` option long before #317, that just polyfilled it for iOS.

**Comment by majornista** [comment]
> I don't follow. Popovers should never extend beyond the viewport. The content of the popover itself is scrollable, not the document body.
> 
> Regardless, closing on scroll has very little to do with this and it is far from the only reason for the current behavior. In fact, this is the first I’m hearing of it in relation to this PR. We used the `preventScroll` option long before #317, that just polyfilled it for iOS.

They should never extend beyond the viewport, except when they do or should; if a user has used pinch and zoom on a mobile device, the contents of an overlay, like ContextualHelp for example, extend beyond the browser viewport, or in the use case @LFDanLu sites above, `shouldFlip:false` and `position: bottom`. `focusWithoutScrolling` is useful in some cases, I just don't think it should replace browser behavior in all cases. I think it is reasonable to have a table or listbox that is taller than the available browser viewport, and in those cases a keyboard users should expect items receiving focus within the table or listbox to scroll into view within both the table or listbox and the browser.

**Comment by devongovett** [comment]
> if a user has used pinch and zoom on a mobile device, the contents of an overlay, like ContextualHelp for example, extend beyond the browser viewport

If that's the case, then it is a bug. If you zoom in a desktop browser, we still keep overlays within the viewport. The behavior should be the same on mobile. Otherwise it's really hard to use if the overlay goes outside the viewport. Definitely not intentional.

> focusWithoutScrolling is useful in some cases, I just don't think it should replace browser behavior in all cases.

Yes, again, `focusWithoutScrolling` is just a polyfill of the `preventScroll` option of the browser `focus` method. You don't want items scrolling under your pointer when using a mouse or touch screen, so it makes sense that this is used. Agree that for keyboard you want items to scroll into view, but I don't think it's as simple as just changing that option. For virtualizer, we have the additional complexity that the item you want to scroll into view might not even be in the DOM, so we cannot simply rely on browser behavior there.

**Comment by majornista** [comment]
@devongovett and @LFDanLu What I'm trying to say is that the `shouldFlip:false` and `position: bottom` use case Daniel describes above would be a broken experience even without the changes in this PR, so it's kind of a moot point.

I don't think of `focusSafely` as the same as `focusWithoutScrolling`, even though it currently uses it. In a virtualizer we don't want mouse focus to cause the browser to scroll a focused element. I understand this. However, I don't think `focusSafely` should completely override a user's expectation for the browser to automatically scroll an item receiving focus into view. The PR is designed to first `focusWithoutScrolling`, and then once the item has received focus without causing jank, and excluding focus after pointer events, verify that the focused element is within view, and if not, scroll it into view.

**Comment by majornista** [comment]
Closing in favor of #3899

**Pull request closed without merging**